src/ApplicationRecordLog.cpp

```diff
@@ -0,0 +1,257 @@
+    }
+
+    void ApplicationRecordLogImp::epoch(const geopm_time_s &time)
+    {
+        check_setup();
+        std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
+        m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
+        check_reset(layout);
+
+        ++m_epoch_count;
+        record_s epoch_record = {
+           .time = geopm_time_diff(&m_time_zero, &time),
+           .process = m_process,
+           .event = EVENT_EPOCH_COUNT,
+           .signal = m_epoch_count,
+        };
+        append_record(layout, epoch_record);
+    }
+
+    void ApplicationRecordLogImp::dump(std::vector<record_s> &records,
+                                       std::vector<short_region_s> &short_regions)
+    {
+        // this function should not do anything with m_hash_region_enter_map
+        std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
+        m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
+        records.resize(layout.num_record);
+        std::copy(layout.record_table, layout.record_table + layout.num_record, records.begin());
+        short_regions.resize(layout.num_region);
+        std::copy(layout.region_table, layout.region_table + layout.num_region, short_regions.begin());
+        layout.num_record = 0;
+        layout.num_region = 0;
+    }
+
+    void ApplicationRecordLogImp::check_setup(void)
+    {
+        if (m_is_setup) {
+            return;
+        }
+
+        if (m_process == -1) {
+            throw Exception("ApplicationRecordLog: set_process() must be called prior to calling enter(), exit() or epoch()",
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        geopm_time_s zero = {{0, 0}};
+        if (geopm_time_diff(&m_time_zero, &zero) == 0.0) {
+            throw Exception("ApplicationRecordLog: set_time_zero() must be called prior to calling enter(), exit() or epoch()",
+                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        m_is_setup = true;
+    }
+
+    void ApplicationRecordLogImp::check_reset(m_layout_s &layout)
+    {
+        if (layout.num_record == 0) {
+            auto region_enter_it = m_hash_region_enter_map.find(m_entered_region_hash);
+            if (region_enter_it != m_hash_region_enter_map.end()) {
+                auto &entry_info = region_enter_it->second;
+                entry_info.record_idx = -1;
+                entry_info.region_idx = -1;
+                m_hash_region_enter_map = {*region_enter_it};
+            }
+            else {
+                m_hash_region_enter_map.clear();
+            }
+        }
+    }
+
+
+    void ApplicationRecordLogImp::append_record(m_layout_s &layout, const record_s &record)
+    {
+        int record_idx = layout.num_record;
+        // Don't overrun the buffer
+        if (record_idx < M_MAX_RECORD) {
+            layout.record_table[record_idx] = record;
```

Comment:
Since record is just being copied anyways, could it be a pass-by-value instead? Then we could do `layout.record_table[record_idx] = std::move(record);`. Compilers should be easily able to perform copy elision for both of the callers' cases, since record goes in and out of scope immediately before/after passing to this function at the callers. Maybe they can even optimize away the reference in the existing implementation, so maybe the net result is the same.

I understand that this code is in a hot path, but maybe this is too micro-optimizey. Might be better left unless we can measure any improvement.